### PR TITLE
Add ability to return TextSelection from SelectableRegion

### PR DESCRIPTION
**Pull request by Dall127**
This PR adds two APIs to Selectable: getLocalTextSelection, and getContentLength. It then overrides these in SelectableRegion to provide the ability to get 'global' offsets for the textSelection. The previous SelectedContent only returned the rawText, whereas this will return the rawText, and conditionally the TextSelection via these 

By doing this, we've created a safe way to override these in _SelectableFragment without exposing any underlying implementation details, as I was told it is unstable at the moment.

This PR  includes unitTests, documentation, and implementation for the described behavior.

Fixes 
 - [Flutter Issue #137371](https://github.com/flutter/flutter/issues/110594)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

@chunhtai Here's the new branch without the formatting issues
(Copied from previous PR)

### CODE REVIEW

File: packages/flutter/lib/src/rendering/paragraph.dart [hunk 1]
```diff
@@ -2071,4 +2071,20 @@ class _SelectableFragment with Selectable, ChangeNotifier implements TextLayoutM
...
+      final int start = math.min(_textSelectionStart!.offset, _textSelectionEnd!.offset);
+      final int end = math.max(_textSelectionStart!.offset, _textSelectionEnd!.offset);
+
+      return TextSelection(baseOffset: start, extentOffset: end);
+    }
+    return null;
+  }
+
+  @override
+  int? getContentLength() {
```
Comment (chunhtai) [inline packages/flutter/lib/src/rendering/paragraph.dart:2087]: What is the use case for this field?

File: packages/flutter/lib/src/widgets/selectable_region.dart [hunk 2]
```diff
@@ -1793,6 +1794,53 @@ abstract class MultiSelectableSelectionContainerDelegate extends SelectionContai
...
+
+  @override
+  TextSelection? getLocalTextSelection() {
+    // initialize data structures
+    int start = 0;
+    int? end;
+    int numSelected = 0;
+    bool enteredSelectedRegion = false;
+
+    for (final Selectable selectable in selectables) {
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/selectable_region.dart:1814]: This should only loop through currentSelectionStartIndex to currentSelectionEndIndex

although you need to consider case where the selection can be inverse

File: packages/flutter/lib/src/rendering/selection.dart [hunk 3]
```diff
@@ -98,6 +98,16 @@ abstract class SelectionHandler implements ValueListenable<SelectionGeometry> {
   /// See also:
   ///  * [SelectionEventType], which contains all of the possible types.
   SelectionResult dispatchSelectionEvent(SelectionEvent event);
+
+  /// Gets the TextSelection from the selected content
+  ///
+  /// Return `null` if nothing is selected, or is not supported by the selectable
+  TextSelection? getLocalTextSelection();
```
Comment (chunhtai) [inline packages/flutter/lib/src/rendering/selection.dart:105]: How would one use this information if there isn't an API to know the content?

**Review by chunhtai — COMMENTED** [review]
It would be good to have an example in flutter/examples/api/lib/material/selection_area to show case how one would use the API to solve a use case.

File: packages/flutter/lib/src/rendering/paragraph.dart (hunk 1 above)
Comment (Dall127) [reply packages/flutter/lib/src/rendering/paragraph.dart:2087]: In the MultiSelectableSelectionContainerDelegate, (where we actually get our global offsets) getContentLength is used method  to add up the lengths of content from each unselected selectable. This helps to figure out where the text selection's index should begin.

File: packages/flutter/lib/src/widgets/selectable_region.dart (hunk 2 above)
Comment (Dall127) [reply packages/flutter/lib/src/widgets/selectable_region.dart:1814]: Since our goal is to determine a global offset, which is crucial in widgets that manage multiple selectables like the MultiSelectableSelectionContainerDelegate, it's important to account for all offsets, not just those that are selected.

File: packages/flutter/lib/src/rendering/selection.dart (hunk 3 above)
Comment (Dall127) [reply packages/flutter/lib/src/rendering/selection.dart:105]: I think this is a valid point, and maybe it does need a function to get the content. At least for my use case (and I assume other's who'd need this type of information) already know's their content and their content's offsets that they're displaying. Perhaps a separate PR if this becomes a need?

**Comment by Dall127** [comment]
> It would be good to have an example in flutter/examples/api/lib/material/selection_area to show case how one would use the API to solve a use case.

@chunhtai while I'm not opposed to making an example in ```flutter/examples/api/lib/material/selection_area ``` as per your suggestion, I would like to reserve that work in the case everything else looks good for this PR. I'll put together a list of use cases from other issues on GitHub in this comment below for now: 

1) When a user selects "pick" in "Peter Piper picked a peck of pickled peppers", there is no way to tell which one of the two "pick"s in the sentence should be highlighted, without necessary info on the selection. This is useful information for fetching definitions 

2) apps like Medium and Kindle, where you can highlight some text and store the highlighted positions so that whenever one opens an article/book/etc. again, the highlight would still be there.

**Comment by Liloupar** [comment]
Can this PR be merged? As previously mentioned, obtaining the index of the selected region is essential for implementing the highlighter feature in the reader.

**Comment by chunhtai** [comment]
> @chunhtai while I'm not opposed to making an example in `flutter/examples/api/lib/material/selection_area ` as per your suggestion, I would like to reserve that work in the case everything else looks good for this PR. I'll put together a list of use cases from other issues on GitHub in this comment below for now:

The reason I ask for example is to see how these new APIs are used. Right now, I am not sure how to make sure of these APIs. but if you are ok with back and forth conversation, then skipping the example for now is fine by me.

> 1. When a user selects "pick" in "Peter Piper picked a peck of pickled peppers", there is no way to tell which one of the two "pick"s in the sentence should be highlighted, without necessary info on the selection. This is useful information for fetching definitions

even if you have text selection offset, you would still need to know the entire content to make sense of it, right?

> 2. apps like Medium and Kindle, where you can highlight some text and store the highlighted positions so that whenever one opens an article/book/etc. again, the highlight would still be there.

global selection is made for just one active selection at a time. It seems to me that this use case is better suited for the search feature that @goderbauer was working on. The API to expose all text in a widget subtree may depend on the outcome of that project.

**Review by chunhtai — COMMENTED** [review]
see my comment above

**Comment by chunhtai** [comment]
> Yes, to make sense of the TextSelection, knowing the entire content is necessary. However, this is typically manageable for developers.

I can understand that some use cases would be ok if we only expose TextSelection. 

The thing I am worry about is that we may at some point need to support cases where the the developers doesn't know the content under selectionArea. This is common if they are developing packages.

That being said, I think exposing the text underneath may be a big project, because we want to make sure whatever we come up with here would support in-page search as well.

**Comment by chunhtai** [comment]
See this doc https://docs.google.com/document/d/1doToFSPGJO29_FJ-VE7mD6XiNgF-s2xVaIi-UC495m8/edit?pli=1&tab=t.0

**Comment by Dall127** [comment]
I'll be updating the branch to pass the tests again, but in the meantime, @chunhtai I think that all looks awesome, and I think there will be some overlap, but the groundwork for getting the information to the users I think is a very solid implementation. Now, I can see that the work being done for actually getting the TextSelection in a cohesive manner changing, but for the time being we're not programming ourselves into a metaphorical corner. We'd be able to swap it out with whatever implementation we want

File: packages/flutter/lib/src/widgets/selectable_region.dart [hunk 4]
```diff
@@ -573,9 +573,10 @@ class SelectableRegionState extends State<SelectableRegion> with TextSelectionDe
   }
 
   void _updateSelectedContentIfNeeded() {
-    if (_lastSelectedContent?.plainText != _selectable?.getSelectedContent()?.plainText) {
+    if (_lastSelectedContent?.plainText !=_selectable?.getSelectedContent()?.plainText) {
       _lastSelectedContent = _selectable?.getSelectedContent();
-      widget.onSelectionChanged?.call(_lastSelectedContent);
+      final TextSelection? textSelection = _selectable?.textSelection;
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/widgets/selectable_region.dart:578]: Will `_selectable?.getSelectedContent().textSelection` also have the global selection points? If so is there a reason why `SelectedContent` and `Selectable` both provide access to the text selection?

File: packages/flutter/lib/src/widgets/selectable_region.dart [hunk 5]
```diff
@@ -1829,6 +1830,53 @@ abstract class MultiSelectableSelectionContainerDelegate extends SelectionContai
     selectable.removeListener(_handleSelectableGeometryChange);
   }
 
+  @override
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/widgets/selectable_region.dart:1833]: should probably also override `getSelectedContent()` so it has the most up to date selection.

File: dev/benchmarks/test_apps/stocks/lib/i18n/stock_strings.dart [hunk 6]
```diff
@@ -14,14 +14,14 @@ import 'package:intl/intl.dart' as intl;
 import 'stock_strings_en.dart';
 import 'stock_strings_es.dart';
 
-/// Callers can lookup localized strings with an instance of StockStrings returned
-/// by `StockStrings.of(context)`.
+/// Callers can lookup localized strings with an instance of StockStrings
```
Comment (Renzo-Olivares) [inline dev/benchmarks/test_apps/stocks/lib/i18n/stock_strings.dart:17]: Are the changes in this directory needed for this PR? If not, then we should revert them.

File: packages/flutter/lib/src/rendering/paragraph.dart [hunk 7]
```diff
@@ -364,6 +364,7 @@ class RenderParagraph extends RenderBox with ContainerRenderObjectMixin<RenderBo
     return results;
   }
 
+
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/rendering/paragraph.dart:367]: nit: is this line needed?

File: packages/flutter/lib/src/rendering/selection.dart [hunk 8]
```diff
@@ -146,6 +171,14 @@ mixin Selectable implements SelectionHandler {
   /// in local coordinates.
   List<Rect> get boundingBoxes;
 
+  /// The [TextSelection] of this [Selectable].
+  @override
+  TextSelection? get textSelection;
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/rendering/selection.dart:176]: Does `Selectable` need to expose this, or can we keep this in `SelectedContent`?

File: packages/flutter/lib/src/widgets/selectable_region.dart [hunk 9]
```diff
@@ -2131,6 +2179,7 @@ abstract class MultiSelectableSelectionContainerDelegate extends SelectionContai
     );
   }
 
+
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/widgets/selectable_region.dart:2182]: nit: consider removing this new line.

File: packages/flutter/lib/src/rendering/selection.dart (hunk 8 above)
Comment (Dall127) [reply packages/flutter/lib/src/rendering/selection.dart:176]: Selectable needs to specifically expose this so that we can get this information at the Selectable's level. When we're building our 'global' selection in routines that use Selectable(s), we know their individual selection

File: examples/api/lib/material/selectable_region/selectable_region.0.dart [hunk 10]
```diff
@@ -270,6 +270,11 @@ class _RenderSelectableAdapter extends RenderProxyBox with Selectable, Selection
     return value.hasSelection ? const SelectedContent(plainText: 'Custom Text') : null;
   }
 
+  @override
+  int? get contentLength => 1;
+
```
[the remark this answers is not in the document]
Comment (Dall127) [reply examples/api/lib/material/selectable_region/selectable_region.0.dart:275]: 👍

File: dev/benchmarks/test_apps/stocks/lib/i18n/stock_strings.dart (hunk 6 above)
Comment (Dall127) [reply dev/benchmarks/test_apps/stocks/lib/i18n/stock_strings.dart:17]: Reverted

File: packages/flutter/lib/src/widgets/selectable_region.dart (hunk 4 above)
Comment (Dall127) [reply packages/flutter/lib/src/widgets/selectable_region.dart:578]: SelectedContent is what is returned to the users; So in the case where we're building the selected content in a different way (like in the MultiSelectionDelegate) we're getting the aggregate of the text_selection of all the selectables. There is a meaningful distinction between the two.

**Comment by Renzo-Olivares** [comment]
Hi @Dall127 thanks for the contribution, just wanted to follow up regarding the comments I left in my previous review. What we are trying to accomplish with this API reminds me of a similar concept on the web `window.getSelection()`. I wonder if there are some learnings we can take from that API in how we choose to represent the data here. WRT to the web, `window.getSelection()` returns a `Selection` object and from that `Selection` object `getRangeAt()` can be used to obtain a `Range` object which includes more details about the current selection such as what node it begins/ends in. I don't think we have to design it exactly like this here, but if there already exists a type of representation for a text selection across multiple nodes/fragments that a user expects then I think it is important to consider it.

**Comment by Dall127** [comment]
@Renzo-Olivares, I can see how that would make more sense in the future; I know in talking with @chunhtai a little bit about it, there was a resounding desire not to expose too many things, the reason being that the underlying APIs are considered to be unstable with their implementations, and are likely to be changed as time goes on (correct me if I'm wrong on any of this @chunhtai ). That being said, I feel as though this PR does an excellent job of balancing the two desires of getting the desired data (it seems like on the linked issue it solves, there's a lot of desire for it) while also keeping implementation & data-structures only accessible in the framework. Even if this only ends up being a stop-gap, I feel like this still sets good groundwork for work towards implementing something akin to the web api you mentioned.

All that being said, @Renzo-Olivares, what are the steps to approve this PR?

**Comment by Renzo-Olivares** [comment]
@Dall127 Apologies for the delay on the review! I will take a look at this PR this week, I think its a reasonable change and the functionality already exists in `SelectableText`. I want to try out the format of the data (`TextSelection`) with the different use-cases that have been given here and in the original issue. Also I think there are some unresolved comments from my last review https://github.com/flutter/flutter/pull/138654#discussion_r1464026993 , https://github.com/flutter/flutter/pull/138654#discussion_r1464028559 , https://github.com/flutter/flutter/pull/138654#discussion_r1464032728 . Thank you!

**Comment by Dall127** [comment]
@Renzo-Olivares that all sounds great, thank you! 

> Also I think there are some unresolved comments from my last review [#138654 (comment)](https://github.com/flutter/flutter/pull/138654#discussion_r1464026993) , [#138654 (comment)](https://github.com/flutter/flutter/pull/138654#discussion_r1464028559) , [#138654 (comment)](https://github.com/flutter/flutter/pull/138654#discussion_r1464032728) . Thank you!

Sorry, I did miss one of them, but the other two I did write responses back. I may not have just properly marked them as being \`resolved\`? In full candidness, this is my first time doing a PR on an open source project, so my GitHub skills are a little lacking in this area. 

Otherwise, let me know if you need anything else from me and I'll try to respond as soon as I can. Thanks for taking a look at it!

File: packages/flutter/lib/src/rendering/selection.dart [hunk 11]
```diff
@@ -98,19 +98,44 @@ abstract class SelectionHandler implements ValueListenable<SelectionGeometry> {
...
-  /// Only supports plain text.
-  const SelectedContent({required this.plainText});
+  /// not all selected content supports TextSelection
+  const SelectedContent({required this.plainText, this.textSelection});
 
   /// The selected content in plain text format.
   final String plainText;
+
+  /// The selected content in [TextSelection] format.
+  final TextSelection? textSelection;
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/rendering/selection.dart:125]: It looks like we aren't initializing any of our `SelectedContent` usages in the framework with a `TextSelection`, is this correct? It looks like we only set this `textSelection` through `copyWith` in `_updateSelectedContentIfNeeded`.

File: packages/flutter/lib/src/rendering/selection.dart [hunk 12]
```diff
@@ -98,19 +98,44 @@ abstract class SelectionHandler implements ValueListenable<SelectionGeometry> {
...
 
 /// The selected content in a [Selectable] or [SelectionHandler].
 // TODO(chunhtai): Add more support for rich content.
 // https://github.com/flutter/flutter/issues/104206.
 class SelectedContent {
   /// Creates a selected content object.
-  ///
-  /// Only supports plain text.
-  const SelectedContent({required this.plainText});
+  /// not all selected content supports TextSelection
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/rendering/selection.dart:118]: This comment should probably be directly above the `textSelection` member.

File: packages/flutter/lib/src/rendering/selection.dart (hunk 11 above)
Comment (Renzo-Olivares) [reply packages/flutter/lib/src/rendering/selection.dart:125]: I think we should be initializing our usages of `SelectedContent` across the framework with a `TextSelection` when its available to us since if a `Selectable` has a non-null `SelectedContent` this means it has a valid selection.

File: packages/flutter/lib/src/rendering/selection.dart (hunk 8 above)
Comment (Renzo-Olivares) [reply packages/flutter/lib/src/rendering/selection.dart:176]: Following up on this comment, it seems to me that this can be contained within `SelectedContent` or am I missing something?

File: packages/flutter/lib/src/rendering/selection.dart [hunk 13]
```diff
@@ -207,6 +240,10 @@ mixin SelectionRegistrant on Selectable {
       _subscribedToSelectionRegistrar = false;
     }
   }
+
+  // not applicable since we don't know the type here
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/rendering/selection.dart:244]: I think we can avoid having this if the `TextSelection` is self-contained within `SelectedContent`.

File: packages/flutter/lib/src/widgets/selectable_region.dart [hunk 14]
```diff
@@ -1883,6 +1884,53 @@ abstract class MultiSelectableSelectionContainerDelegate extends SelectionContai
...
+  int? get contentLength {
+    int length = 0;
+    for (final Selectable selectable in selectables) {
+      length += selectable.contentLength ?? 0;
+    }
+    return length;
+  }
+
+  @override
+  TextSelection? get textSelection {
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/widgets/selectable_region.dart:1897]: Can this calculation go in the override of `getSelectedContent`? This way the `SelectedContent` have the most up to date `TextSelection` and `_updateSelectedContentIfNeeded` will only need to call `getSelectedContent()` rather than having to access `_selectable.textSelection`. Let me know if this is not possible, or if I can provide a better explanation.

File: packages/flutter/lib/src/widgets/selectable_region.dart (hunk 4 above)
Comment (Renzo-Olivares) [reply packages/flutter/lib/src/widgets/selectable_region.dart:608]: Following up on this comment.

File: packages/flutter/lib/src/widgets/selectable_region.dart [hunk 15]
```diff
@@ -1883,6 +1884,53 @@ abstract class MultiSelectableSelectionContainerDelegate extends SelectionContai
...
+    int start = 0;
+    int? end;
+    int numSelected = 0;
+    bool enteredSelectedRegion = false;
+
+    for (final Selectable selectable in selectables) {
+      // Retrieve the selected content, if any, of the current 'selectable'.
+      final SelectedContent? data = selectable.getSelectedContent();
+      if (data != null) {
+        final TextSelection? selection = selectable.textSelection ;
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/widgets/selectable_region.dart:1908]: nit: unnecessary whitespace before the semi-colon.

**Comment by Dall127** [comment]
@Renzo-Olivares I saw what you mean with rolling `TextSelection` into `SelectedContent`, and went ahead and did that. The only thing that a `Selectable` exposes now, after this change, is `contentLength`. This one is still needed, since regardless of whether or not it's selected, we still need this information from a `Selectable` for building a global offset.

**Comment by Renzo-Olivares** [comment]
Hi @Dall127, thank you for your contribution. I apologize for the delay on this review and appreciate your patience with this process.

There has been a recent motivation to develop a more robust API to expose the current selection information, which warrants closing this one. This decision is motivated by the goal to accommodate a wider range of use cases than the current `TextSelection` format allows.

Once the new PR is up, I'll make sure to link the issue, and your insights or feedback on the feature would be highly appreciated. Your work on this PR has been invaluable, and we're grateful for your involvement.

**Comment by Liloupar** [comment]
@Renzo-Olivares Is there an approximate release time for the new PR? Thank you.

**Comment by Liloupar** [comment]
> Hi @Dall127, thank you for your contribution. I apologize for the delay on this review and appreciate your patience with this process.
> 
> There has been a recent motivation to develop a more robust API to expose the current selection information, which warrants closing this one. This decision is motivated by the goal to accommodate a wider range of use cases than the current `TextSelection` format allows.
> 
> Once the new PR is up, I'll make sure to link the issue, and your insights or feedback on the feature would be highly appreciated. Your work on this PR has been invaluable, and we're grateful for your involvement.

@Renzo-Olivares is any update for this? thanks

**Comment by nathanael-h** [comment]
https://github.com/flutter/flutter/pull/148998
